authorization-server/src/modules/auth/auth.controller.ts

```diff
@@ -189,24 +191,70 @@ export class AuthController {
   @UseGuards(ValidRefreshTokenGuard)
   @ApiBody({ type: RefreshDto })
   @ApiOkResponse({ type: LoginResponseDto })
-  async refresh(
+  // backwards compatibility
+  async refreshWithPost(
     @Body() body: RefreshDto,
     @Res({ passthrough: true }) res: Response,
   ): Promise<LoginResponseDto | undefined> {
-    const { accessToken, refreshToken } = await this.authService.refreshTokens(
-      body.refreshToken,
+    return this.refreshCommon(body.refreshToken, res);
+  }
+
+  @Get('refresh_token')
+  //compatibile with the SSI-HUB implementation
+  async refreshWithGet(
+    @Req() req: Request,
+    @Res({ passthrough: true }) res: Response,
+    @Query('refresh_token') refreshTokenFromQueryString: string,
+  ): Promise<LoginResponseDto | undefined> {
+    let refreshToken: string;
+
+    if (this.configService.get<boolean>('AUTH_COOKIE_ENABLED')) {
+      refreshToken =
+        refreshTokenFromQueryString ||
+        (req.cookies || {})[
+          this.configService.get('AUTH_COOKIE_NAME_REFRESH_TOKEN')
+        ];
+    } else {
+      refreshToken = refreshTokenFromQueryString;
+    }
+
+    if (!refreshToken) {
+      this.logger.warn(`no refresh token provided`);
+      throw new UnauthorizedException();
+    }
+
+    const refreshTokenIsValid = await this.authService.validateRefreshToken(
```

Comment:
Yes, I was thinking about this. Having only guards makes writing tests simpler because logic is isolated. But... having a shared guard would mix logic for POST and GET implementations.

Maybe we could have two separate guards for POST and GET?

Reply:
Yes, I think having two separate guards could make sense 👍

Reply:
I've just noticed a bug in the `refreshWithPost()`. It gets a token only from the request body. So, it looks like quite a lot of logic from the guard needs to be repeated to make this work as expected. Or a token can be attached to the request object in the guard and then used in the method.

Reply:
In my opinion token verification is the same for both endpoints. These endpoints differs only how refresh token is extracted. Would it make sense to have one guard and two pipes? I think the pipe is more appropriate component for extracting then guard

Reply:
This could a good idea to extract and verify tokens separately but looks like pipes are executed after guards: https://docs.nestjs.com/faq/request-lifecycle. I need to do some research anyway.

Reply:
Right, then another option is to refuse from guard in favor of `verify` method. And the only difference between GET and POST handlers would be which of cookie, query or body extractor pipes are applied

Reply:
> Right, then another option is to refuse from guard in favor of `verify` method

Yes, I think this could work too. I think symmetry would be ideal (i.e. both using guards or both using pipes or both using class functions, etc.)

Reply:
I think it could be solved by the guard storing the refresh token in the `request.user` that will be used in methods. Extraction conditions would depend on the request method - `GET` or `POST`.

Reply:
It seems to me that guard should not be allowed to transform request. Its single responsibility is to decide whether request is allowed or not

Reply:
actually, guards do this in case of login strategies

Reply:
maybe we should just have a token extraction method used both in guard and then in endpoint methods?

Reply:
I am getting convinced about storing tokens as `req.user`. I see that both login and jwt strategies do this.

Reply:
> maybe we should just have a token extraction method used both in guard and then in endpoint methods?

Makes sense to me 👍

Reply:
> I am getting convinced about storing tokens as req.user. I see that both login and jwt strategies do this.

I think maybe I see @JGiter 's intuition that transforming requests is not ideal for guards, but I think if we look at `req.user` as being "what is known about the user so far for this request", I think storing a token in `req.user` could be reasonable.

I don't really understand the `req.user` property though. I doesn't seem like it is a standard property on the request object for express https://expressjs.com/en/api.html#req . Is it just property that the `jwt` strategy decided to add?